### ISSUE TITLE
Add support for forwarding raw packets over TUN interface

### ISSUE THREAD

**Issue by pmeenan** [Genymobile/gnirehtet#110]
Instead of intercepting all packets and proxying the traffic, when root is available on systems with TUN support the packets can be forwarded directly.  That is basically merging the functionality from the [vpn-reverse-tether](https://github.com/WPO-Foundation/vpn-reverse-tether) relay into the gnirehtet relay.

FWIW, I'm planning on implementing this in my fork in the Rust relay for WebPageTest to use with it's agents (where passing raw packets matters).  I'm happy to submit a PR if you are interested in it as part of the main project.

**Comment by rom1v** [comment]
> Instead of intercepting all packets and proxying the traffic, when root is available on systems with TUN support the packets can be forwarded directly.

Yes :) (see https://medium.com/genymobile/gnirehtet-reverse-tethering-android-2afacdbdaec7)

>  That is basically merging the functionality from the [vpn-reverse-tether](https://github.com/WPO-Foundation/vpn-reverse-tether) relay into the gnirehtet relay.

To be sure to understand, why not using vpn-reverse-tether (or [SimpleRT](https://github.com/vvviperrr/SimpleRT)) directly?

**Comment by pmeenan** [comment]
The main reason is that vpn-reverse-tether and SimpleRT have largely been abandoned and the android apps don't work on Nougat or later.  I've considered updating the VPN app on vpn-reverse-tether instead but long-term that feels like a losing proposition.

In my use case specifically I need the packets from android to go through to the origins so that I'm exercising the device's TCP stack (congestion control, etc) and not the host's.

My current work is a hybrid that uses a modified forwarder from vpn-reverse-tether that can talk to the gnirehtet app but folding the functionality into the Rust implementation is also doable (not sure I want to deal with Java).

**Comment by rom1v** [comment]
> I'm happy to submit a PR if you are interested in it as part of the main project.

How would you expose it? (a parameter to `gnirehtet relay` that would use TUN interface?)
Would it imply to run gnirehtet as root? Would it still support multiple clients for the same relay server?

**Comment by pmeenan** [comment]
Yes, it would be an additional parameter (like specifying dns or routes) and would require running gnirehtet under sudo.  Initial release will probably only support a single client but multiple tun interfaces is certainly possible.

Right now I'm experimenting with replacing the forwarding code on the device side with the native code from vpn-reverse-tether to see if that improves performance.  I'm also changing the framing of the tunneled data to include a header with the packet size which will both significantly simplify the processing as well as make a server-side tun forwarder a trivial matter.

Given the number of changes in flight it probably makes sense to just make all of the changes in my fork and see what works out and what doesn't.
